### PR TITLE
Feature/action validation

### PR DESCRIPTION
**Pull request by kkirsz**
I have added the possibility to validate workflow actions (see image below).
photo1
It opens a modal window where a user can select which fields will be validated before workflow action will take place.
photo2
I based this feature on `Pimcore\Model\DataObject\ClassDefinition\Data::checkValidity()` method. Validation rules are saved to the standard `workflowmanagement.php` file, for instance:
```
"actions" => [
            [
                "name" => "makeAccepted",
                "label" => "makeAccepted",
                "transitionTo" => [
                    "publication" => [
                        "accepted"
                    ]
                ],
                "notes" => [
                    "required" => FALSE,
                    "title" => "",
                    "type" => ""
                ],
                "validation" => [
                    [
                        "classId" => "1",
                        "rules" => [
                            "events.date",
                            "localizedfields.name"
                        ]
                    ]
                ],
                "additionalFields" => [

                ],
                "users" => [

                ],
                "notificationUsers" => [

                ],
                "events" => [
                    "before" => [
                        "before"
                    ],
                    "success" => [
                        "success"
                    ],
                    "failure" => [
                        "failure"
                    ]
                ]
            ]
```
I know that Pimcore 5.5 provides a new workflow engine, but this feature still can be useful for older versions. I hope you enjoy this.

### CODE REVIEW

File: src/WorkflowGuiBundle/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<?php
+/**
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/DependencyInjection/Configuration.php]: please add my header here

File: src/WorkflowGuiBundle/DependencyInjection/WorkflowGuiExtension.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
+<?php
+/**
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/DependencyInjection/WorkflowGuiExtension.php:2]: please add my header here

File: src/WorkflowGuiBundle/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+namespace WorkflowGuiBundle\DependencyInjection;
+
+use Symfony\Component\Config\Definition\Builder\TreeBuilder;
+use Symfony\Component\Config\Definition\ConfigurationInterface;
+
+/**
+ * Class Configuration
+ * @package WorkflowGuiBundle\DependencyInjection
+ */
+class Configuration implements ConfigurationInterface
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/DependencyInjection/Configuration.php]: really needed? there is no configuration at all

File: src/WorkflowGuiBundle/EventListener/WorkflowListener.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace WorkflowGuiBundle\EventListener;
+
+use Pimcore\Event\Model\WorkflowEvent;
+use Pimcore\Event\WorkflowEvents;
+use Pimcore\Model\DataObject\Concrete;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use WorkflowGuiBundle\Validation\ValidationManager;
+
+/**
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/EventListener/WorkflowListener.php]: please remove class header

File: src/WorkflowGuiBundle/EventListener/WorkflowListener.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
+<?php
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/EventListener/WorkflowListener.php:1]: please add my header here

File: src/WorkflowGuiBundle/EventListener/WorkflowListener.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ */
+class WorkflowListener implements EventSubscriberInterface
+{
+    /**
+     * @var ValidationManager
+     */
+    protected $validationManager;
+
+    /**
+     * WorkflowListener constructor.
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/EventListener/WorkflowListener.php]: please remove `WorkflowListener constructor` line

File: src/WorkflowGuiBundle/Resources/public/js/workflow/validation.js [hunk 7]
```diff
@@ -0,0 +1,296 @@
+pimcore.registerNS("pimcore.plugin.workflowgui.validation");
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Resources/public/js/workflow/validation.js]: please add headers here

File: src/WorkflowGuiBundle/Validation/DependencyInjection/ServiceLocator.php [hunk 8]
```diff
@@ -0,0 +1,35 @@
+<?php
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/DependencyInjection/ServiceLocator.php:1]: please add my header here

File: src/WorkflowGuiBundle/Validation/DependencyInjection/ServiceLocatorAwareInterface.php [hunk 9]
```diff
@@ -0,0 +1,28 @@
+<?php
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/DependencyInjection/ServiceLocatorAwareInterface.php:1]: please add my header here

File: src/WorkflowGuiBundle/Validation/DependencyInjection/ServiceLocatorAwareTrait.php [hunk 10]
```diff
@@ -0,0 +1,39 @@
+<?php
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/DependencyInjection/ServiceLocatorAwareTrait.php:1]: please add my header here

File: src/WorkflowGuiBundle/Validation/Handler/BlockHandler.php [hunk 11]
```diff
@@ -0,0 +1,104 @@
+<?php
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/Handler/BlockHandler.php:1]: please add my header here

File: src/WorkflowGuiBundle/Validation/Handler/BlockHandler.php [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+declare(strict_types=1);
+
+namespace WorkflowGuiBundle\Validation\Handler;
+
+use Pimcore\Model\DataObject\ClassDefinition\Data;
+use Pimcore\Model\DataObject\Data\BlockElement;
+use WorkflowGuiBundle\Validation\DependencyInjection;
+use WorkflowGuiBundle\Validation\ValidationErrorsTrait;
+
+/**
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/Handler/BlockHandler.php]: please remove class header

File: src/WorkflowGuiBundle/Validation/Handler/GeneralHandler.php [hunk 13]
```diff
@@ -0,0 +1,68 @@
+<?php
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/Handler/GeneralHandler.php:1]: please add my header here

File: src/WorkflowGuiBundle/Validation/Handler/GeneralHandler.php [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+namespace WorkflowGuiBundle\Validation\Handler;
+
+use Pimcore\Model\DataObject\ClassDefinition\Data;
+use Pimcore\Model\Element\ValidationException;
+use WorkflowGuiBundle\Validation\ValidationErrorsTrait;
+
+/**
+ * Class GeneralHandler
+ * @package WorkflowGuiBundle\Validation\Handler
+ */
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/Handler/GeneralHandler.php]: please remove class header

File: src/WorkflowGuiBundle/Validation/Handler/HandlerInterface.php [hunk 15]
```diff
@@ -0,0 +1,33 @@
+<?php
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/Handler/HandlerInterface.php:1]: please add my header here

File: src/WorkflowGuiBundle/Validation/Handler/HandlerInterface.php [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+declare(strict_types=1);
+
+namespace WorkflowGuiBundle\Validation\Handler;
+
+use Pimcore\Model\DataObject\ClassDefinition\Data;
+
+/**
+ * Interface HandlerInterface
+ * @package WorkflowGuiBundle\Validation\Handler
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/Handler/HandlerInterface.php]: please remove class header

File: src/WorkflowGuiBundle/Validation/Handler/LocalizedfieldsHandler.php [hunk 17]
```diff
@@ -0,0 +1,108 @@
+<?php
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/Handler/LocalizedfieldsHandler.php:1]: please add my header here

File: src/WorkflowGuiBundle/Validation/Handler/LocalizedfieldsHandler.php [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+namespace WorkflowGuiBundle\Validation\Handler;
+
+use Pimcore\Model\DataObject\ClassDefinition\Data;
+use Pimcore\Model\DataObject\Localizedfield;
+use WorkflowGuiBundle\Validation\DependencyInjection;
+use WorkflowGuiBundle\Validation\ValidationErrorsTrait;
+
+/**
+ * Class LocalizedfieldsHandler
+ * @package WorkflowGuiBundle\Validation\Handler
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/Handler/LocalizedfieldsHandler.php]: please remove class header

File: src/WorkflowGuiBundle/Validation/ValidationErrorsTrait.php [hunk 19]
```diff
@@ -0,0 +1,59 @@
+<?php
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/ValidationErrorsTrait.php:1]: please add my header here

File: src/WorkflowGuiBundle/Validation/ValidationErrorsTrait.php [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+ * @copyright   Copyright (c) 2018 Divante Ltd. (https://divante.co)
+ */
+
+declare(strict_types=1);
+
+namespace WorkflowGuiBundle\Validation;
+
+/**
+ * Trait ValidationErrorsTrait
+ * @package WorkflowGuiBundle\Validation
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/ValidationErrorsTrait.php]: please remove class headers

File: src/WorkflowGuiBundle/Validation/ValidationManager.php [hunk 21]
```diff
@@ -0,0 +1,173 @@
+<?php
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/ValidationManager.php:1]: please add my header here

File: src/WorkflowGuiBundle/Validation/ValidationManager.php [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+namespace WorkflowGuiBundle\Validation;
+
+use Pimcore\Model\DataObject\AbstractObject;
+use Pimcore\Model\DataObject\ClassDefinition;
+use Pimcore\Model\DataObject\Concrete;
+use Pimcore\WorkflowManagement\Workflow\Manager as WorkflowManager;
+
+/**
+ * Class ValidationManager
+ * @package WorkflowGuiBundle\Validation
```
Comment (dpfaffenbauer) [inline src/WorkflowGuiBundle/Validation/ValidationManager.php]: please remove class header

**Review by dpfaffenbauer — CHANGES_REQUESTED** [review]
AWESOME 👍 +

**Comment by dpfaffenbauer** [comment]
an amazing feature!!!

Sad that the Workflow GUI doesn't support the new Workflow

File: src/WorkflowGuiBundle/DependencyInjection/Configuration.php (hunk 1 above)
Comment (kkirsz) [reply src/WorkflowGuiBundle/DependencyInjection/Configuration.php]: I see you don't use any header in the WorkflowGUI bundle. I understand that no header should be placed?

Comment (dpfaffenbauer) [reply src/WorkflowGuiBundle/DependencyInjection/Configuration.php]: Oh... thats not good :D, Ok. Lets do your PR without headers, and I will add them afterwards

Comment (kkirsz) [reply src/WorkflowGuiBundle/DependencyInjection/Configuration.php]: Do you mean header from `src/WorkflowGuiBundle/Resources/public/js/workflow/item.js` for instance?

Comment (dpfaffenbauer) [reply src/WorkflowGuiBundle/DependencyInjection/Configuration.php]: yes, that would be the right one.

File: src/WorkflowGuiBundle/DependencyInjection/WorkflowGuiExtension.php (hunk 2 above)
Comment (kkirsz) [reply src/WorkflowGuiBundle/DependencyInjection/WorkflowGuiExtension.php:2]: done

File: src/WorkflowGuiBundle/EventListener/WorkflowListener.php (hunk 5 above)
Comment (kkirsz) [reply src/WorkflowGuiBundle/EventListener/WorkflowListener.php:1]: done

File: src/WorkflowGuiBundle/Validation/DependencyInjection/ServiceLocator.php (hunk 8 above)
Comment (kkirsz) [reply src/WorkflowGuiBundle/Validation/DependencyInjection/ServiceLocator.php:1]: done

File: src/WorkflowGuiBundle/Validation/DependencyInjection/ServiceLocatorAwareInterface.php (hunk 9 above)
Comment (kkirsz) [reply src/WorkflowGuiBundle/Validation/DependencyInjection/ServiceLocatorAwareInterface.php:1]: done

File: src/WorkflowGuiBundle/Validation/DependencyInjection/ServiceLocatorAwareTrait.php (hunk 10 above)
Comment (kkirsz) [reply src/WorkflowGuiBundle/Validation/DependencyInjection/ServiceLocatorAwareTrait.php:1]: done

File: src/WorkflowGuiBundle/Validation/Handler/BlockHandler.php (hunk 11 above)
Comment (kkirsz) [reply src/WorkflowGuiBundle/Validation/Handler/BlockHandler.php:1]: done

File: src/WorkflowGuiBundle/Validation/Handler/GeneralHandler.php (hunk 13 above)
Comment (kkirsz) [reply src/WorkflowGuiBundle/Validation/Handler/GeneralHandler.php:1]: done

File: src/WorkflowGuiBundle/Validation/Handler/HandlerInterface.php (hunk 15 above)
Comment (kkirsz) [reply src/WorkflowGuiBundle/Validation/Handler/HandlerInterface.php:1]: done

File: src/WorkflowGuiBundle/Validation/Handler/LocalizedfieldsHandler.php (hunk 17 above)
Comment (kkirsz) [reply src/WorkflowGuiBundle/Validation/Handler/LocalizedfieldsHandler.php:1]: done

File: src/WorkflowGuiBundle/Validation/ValidationErrorsTrait.php (hunk 19 above)
Comment (kkirsz) [reply src/WorkflowGuiBundle/Validation/ValidationErrorsTrait.php:1]: done

File: src/WorkflowGuiBundle/Validation/ValidationManager.php (hunk 21 above)
Comment (kkirsz) [reply src/WorkflowGuiBundle/Validation/ValidationManager.php:1]: done

File: composer.json [hunk 23]
```diff
@@ -14,6 +14,12 @@
       "email": "dominik@pfaffenbauer.at",
       "homepage": "https://www.pfaffenbauer.at/",
       "role": "Developer"
+    },
+    {
+      "name": "Korneliusz Kirsz",
```
Comment (dpfaffenbauer) [inline composer.json]: not sure about that ;)

I think we should add the community/contributors with one entry: https://github.com/symfony/symfony/blob/master/composer.json#L14
https://github.com/Sylius/Sylius/blob/master/composer.json#L17

**Review by dpfaffenbauer — CHANGES_REQUESTED** [review]

File: composer.json (hunk 23 above)
Comment (kkirsz) [reply composer.json]: done

**Comment by kkirsz** [comment]
As you can see, I have fixed everything you suggested. Do you think PR is OK now?

**Comment by dpfaffenbauer** [comment]
no need for a hurry, I need to test everything before merging it ;)

**Comment by dpfaffenbauer** [comment]
alright, looks all good to me :)

thanks a lot for this PR, and maybe together we are able to push this plugin for the new Workflow Engine :)

**Review by dpfaffenbauer — APPROVED** [review]

**Comment by kkirsz** [comment]
Thanks a lot. I look forward to the new workflow engine :)

**Pull request merged**